sklearn/ensemble/iforest.py

```diff
@@ -0,0 +1,268 @@
+        n_samples = X.shape[0]
+        if max_samples > n_samples:
+            warn("max_samples (%s) is greater than the "
+                 "total number of samples (%s). max_samples "
+                 "will be set to n_samples for estimation."
+                 % (self.max_samples, n_samples))
+            max_samples = n_samples
+
+        super(IsolationForest, self)._fit(X, y, max_samples,
+                                          sample_weight=sample_weight)
+        return self
+
+    def predict(self, X):
+        """Predict anomaly score of X with the IsolationForest algorithm.
+
+        The anomaly score of an input sample is computed as
+        the mean anomaly scores of the trees in the forest.
+
+        The measure of normality of an observation given a tree is the depth
+        of the leaf containing this observation, which is equivalent to
+        the number of splitting required to isolate this point. In case of
+        several observations n_left in the leaf, the average length path of
+        a n_left samples isolation tree is added.
+
+        Parameters
+        ----------
+        X : array-like or sparse matrix of shape (n_samples, n_features)
+            The input samples. Internally, it will be converted to
+            ``dtype=np.float32`` and if a sparse matrix is provided
+            to a sparse ``csr_matrix``.
+
+        Returns
+        -------
+        scores : array of shape (n_samples,)
+            The anomaly score of the input samples.
+            The lower, the more normal.
+        """
+        # code structure from ForestClassifier/predict_proba
+        # Check data
+        X = self.estimators_[0]._validate_X_predict(X, check_input=True)
+        n_samples = X.shape[0]
+
+        # Assign chunk of trees to jobs
+        n_jobs, n_trees, starts = _partition_estimators(self.n_estimators,
+                                                        self.n_jobs)
+
+        # Parallel loop
+        results = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,
+                           backend="threading")(
+                               delayed(_parallel_helper)(tree.tree_,
+                                                         'apply_depth', X)
+                               for tree in self.estimators_)
+
+        # We had to add to the depth (results) the average depth wrt to
+        # the number of samples remaining in the leaf: First find this number:
+        n_samples_leaf = np.zeros((self.n_estimators, n_samples))
+        estimator_nb = -1
+        for tree in self.estimators_:
+            estimator_nb += 1
+            leaves_index = tree.apply(X)
+            n_nodes = tree.tree_.n_node_samples
+            n_samples_leaf[estimator_nb, :] = n_nodes[leaves_index]
+
+        # Now, the depth if the tree was fully grown: (add an average term)
+        depth_if_full = np.array(results) \
+                        + _average_path_length(np.array(n_samples_leaf))
```

Comment:
can you avoid using `\`? this is considered as bad practice for various reasons.